File: 114-flatten-binary-tree-to-linked-list/114-flatten-binary-tree-to-linked-list.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public void flatten(TreeNode root) {
        TreeNode n = root;
        while(n != null) {
            if(n.left == null) {
                n = n.right;
                continue;
            }
            else {
                TreeNode pre = n.left;
                if(n.right != null) {
                    while(pre.left != null || pre.right != null) {
                        if(pre.right != null) {
                            pre = pre.right;
                        }
                        else if(pre.left != null) {
                            pre = pre.left;
                        }
                    }
                    pre.right = n.right;
                }
                n.right = n.left;
                n.left = null;
            }
        }
    }
}

File: 114-flatten-binary-tree-to-linked-list/114-flatten-binary-tree-to-linked-list.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def flatten(self, root: Optional[TreeNode]) -> None:
        stack = []
        curr = root
        prev = None

        while curr or stack:
            while curr:
                if prev:
                    prev.right = curr
                    prev.left = None
                if curr.right:
                    stack.append(curr.right)
                prev = curr
                curr = curr.left
            curr = stack.pop() if stack else None
        